python/ray/train/_internal/dataset_spec.py

```diff
@@ -188,17 +189,34 @@ def get_dataset_shards(
                 )
                 dataset = dataset.window(bytes_per_window=stream_window_size).repeat()
                 # In windowed mode, we re-apply the preprocessor on each iteration.
-                if self.preprocessor:
+                if self.preprocessor or config.per_epoch_preprocessor:
+                    if self.preprocessor is not None:
+                        preprocessor = self.preprocessor
+                        if config.per_epoch_preprocessor is not None:
+                            preprocessor = Chain(
+                                preprocessor, config.per_epoch_preprocessor
+                            )
+                    else:
+                        preprocessor = config.per_epoch_preprocessor
+
                     # TODO: Replace with self.preprocessor.transform when possible.
-                    prep = self.preprocessor.transform_batch
+                    prep = preprocessor.transform_batch
                     dataset = dataset.map_batches(prep, batch_format="pandas")
+
                 # Always re-randomize each window; this doesn't help with reducing
                 # cluster hot-spots since we already randomized the based blocks, but
                 # can help with improving randomness in combination with local shuffle.
                 if config.randomize_block_order and not config.global_shuffle:
                     # TODO(swang): Should randomize block order across the
                     # original dataset, not the window.
                     dataset = dataset.randomize_block_order_each_window()
+            if config.per_epoch_preprocessor is not None:
+                # Reapply the per epoch preprocessor on each epoch.
+                if isinstance(dataset, Dataset):
+                    dataset = dataset.repeat()
+                # TODO: Replace with preprocessor.transform when possible.
+                per_epoch_prep = config.per_epoch_preprocessor.transform_batch
+                dataset = dataset.map_batches(per_epoch_prep)
```

Comment:
Not sure if I'm just being dumb here, but doesn't this apply the per-epoch preprocessor twice on the first epoch? Like, if `config.per_epoch_preprocessor` isn't `None`, then we apply it on both line 204 and line 219?

Reply:
that's a good point...I think lines 218-219 need to be moved inside the if statement.

Reply:
Ah thanks you're right, it should be under an `elif`. I'll add a test to make sure we're only applying it once.